src/dsl/lb.rs

```diff
@@ -0,0 +1,199 @@
+    /// mutable reference to the `LookupBuilder<F>`, it can an chain multiple `add` and `enable`
+    /// function calls to build the lookup table. Requires calling `lookup` to create an empty
+    /// `LookupBuilder` instance at the very front.
+    pub fn add<C: Into<Constraint<F>>, E: Into<Expr<F>>>(
+        mut self,
+        constraint: C,
+        expression: E,
+    ) -> Self {
+        let constraint = constraint.into();
+        self.lookup
+            .add(constraint.annotation, constraint.expr, expression.into());
+        self
+    }
+
+    /// Adds a selector column specific to the lookup table. Because the function returns a mutable
+    /// reference to the `LookupBuilder<F>`, it can an chain multiple `add` and `enable` function
+    /// calls to build the lookup table. Requires calling `lookup` to create an
+    /// empty `LookupBuilder` instance at the very front.
+    pub fn enable<C: Into<Constraint<F>>>(mut self, enable: C) -> Self {
+        let enable = enable.into();
+        self.lookup.enable(enable.annotation, enable.expr);
+        self
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct LookupTableStore<F> {
+    id: UUID,
+    dest: Vec<Expr<F>>,
+}
+
+impl<F> Default for LookupTableStore<F> {
+    fn default() -> Self {
+        Self {
+            id: uuid(),
+            dest: Default::default(),
+        }
+    }
+}
+
+impl<F> LookupTableStore<F> {
+    #[allow(clippy::should_implement_trait)]
+    pub fn add<E: Into<Expr<F>>>(mut self, expr: E) -> Self {
+        self.dest.push(expr.into());
+
+        self
+    }
+
+    pub fn uuid(&self) -> UUID {
+        self.id
+    }
+}
+
+impl<F: Debug + Clone> LookupTableStore<F> {
+    pub fn build(self, src: Vec<Constraint<F>>, enable: Option<Constraint<F>>) -> Lookup<F> {
+        assert_eq!(
+            self.dest.len(),
+            src.len(),
+            "number of source and destination parameters doesn't match"
+        );
+
+        let mut lookup = Lookup::default();
+
+        if let Some(enable) = enable {
+            lookup.enable(enable.annotation, enable.expr);
+        }
+
+        src.into_iter()
+            .zip(self.dest)
+            .for_each(|(src, dest)| lookup.add(src.annotation, src.expr, dest));
+
+        lookup
+    }
+}
+
+#[derive(Debug)]
+pub struct LookupTableRegistry<F>(Arc<Mutex<HashMap<UUID, LookupTableStore<F>>>>);
+
+impl<F> Clone for LookupTableRegistry<F> {
+    fn clone(&self) -> Self {
+        Self(Arc::clone(&self.0))
+    }
+}
+
+impl<F> Default for LookupTableRegistry<F> {
+    fn default() -> Self {
+        Self(Arc::new(Mutex::new(HashMap::new())))
+    }
+}
+
+impl<F> LookupTableRegistry<F> {
+    pub fn add(&self, table: LookupTableStore<F>) {
+        self.0.lock().as_mut().unwrap().insert(table.uuid(), table);
+    }
+}
+
+impl<F: Clone> LookupTableRegistry<F> {
+    pub fn get(&self, uuid: UUID) -> LookupTableStore<F> {
+        (*self.0.lock().unwrap().get(&uuid).unwrap()).clone()
+    }
+}
```

Comment:
Just to clear my understanding, Arc<Mutex<>> is how we create a persistent variable for the super circuit? By persistent I mean it's not globally static, but will persist as long as we have one reference to it?
1. Are we using Arc instead of Rc so that it's thread-safe, i.e. multiple sub-circuits (threads) can access the same LookupRegistry?
2. Why use Mutex instead of RefCells?
3. Why is lazy static not needed? Is it because Arc will persist as long as there are more than one reference, so that we don't need a globally static variable?

Reply:
I'm thinking about using a similar design for PyChiquito Rust end, which stores HashMap<UUID, AST>, where UUID is shared across Python and Rust.

Reply:
For 1), I don't think Rc is thread-safe, so it makes sense to use Arc if we are using multiple threads for the LookupRegistry. 

For 2), I think we need mutexes because the data in the LookupTableRegistry is being updated. I'm not sure about the breadth of their differences, but I know mutexes are usefule for safely managing threads that update state.

For 3), I am not sure, since I can see how using `lazy_static!` could make sense.

Of course, need confirmation from @leolara on these thoughts :)

Reply:
It is an easy way to make a shared object writable (Rc only allows reads), and has the aditional benefit of being multi-thread, although we are not running it multithread now.

Reply:
lazy_static is for global things, this is only at the super_circuit level

Reply:
Got it thanks both!